libsolutil/CDCL.h

```diff
@@ -48,16 +49,137 @@ struct Literal
 		return std::make_tuple(positive, variable) < std::make_tuple(_other.positive, _other.variable);
 	}
 };
-using Clause = std::vector<Literal>;
+
+struct Clause {
+	const Literal& front() const
+	{
+		return lits.front();
+	}
+	const Literal& back() const
+	{
+		return lits.back();
+	}
+	Literal& front()
+	{
+		return lits.front();
+	}
+	Literal& back()
+	{
+		return lits.back();
+	}
+	const Literal& operator[](const size_t at) const
+	{
+		return lits.at(at);
+	}
+	const Literal& at(const size_t at) const
+	{
+		return lits.at(at);
+	}
+	Literal& operator[](const size_t at)
+	{
+		return lits.at(at);
+	}
+	Literal& at(const size_t at)
+	{
+		return lits.at(at);
+	}
+	auto size() const
+	{
+		return lits.size();
+	}
+	auto empty() const
+	{
+		return lits.empty();
+	}
+	auto begin() {
+		return lits.begin();
+	}
+	auto end() {
+		return lits.end();
+	}
+	auto begin() const {
+		return lits.begin();
+	}
+	auto end() const {
+		return lits.end();
+	}
+	void push_back(const Literal& l)
+	{
+		lits.push_back(l);
+	}
+
+	std::vector<Literal> lits;
+	uint64_t ID;
+};
+
+class TriState {
+public:
+	TriState(const bool b)
+	{
+		if (b)
+			val = 1;
+		else
+			val = 0;
+	}
+
+	TriState()
+	{}
+
+	bool operator==(const TriState& other) const
+	{
+		return val == other.val;
+	}
+
+	bool operator!=(const TriState& other) const
+	{
+		return val != other.val;
+	}
+
+	bool toBool() const {
+		if (val == 1)
+			return true;
+		else if (val == 0)
+			return false;
+		else
+		{
+			assert(val== 2);
+			assert(false && "UNSET cannot be converted to bool");
+		}
+	}
+
+	std::string toString()
+	{
+		if (val == 1)
+			return "true";
+		else if (val == 0)
+			return "false";
+		else
+		{
+			assert(val == 2);
+			return "unset";
+		}
+	}
+
+
+private:
+	// Default value is UNSET
+	char val = 2;
```

Comment:
```suggestion
	uint8_t val = 2;
```
(char can be a very weird thing depending on the architecture)

Reply:
Yes, indeed. Actually, the worst part about `char*` is that it's always assumed that [it can alias](https://stackoverflow.com/questions/23848188/strict-aliasing-rule-and-char-pointers) to another pointer type so the compiler cannot use some optimizations. I wanted to use `char` just not to confuse, but `uint8_t` is what I would have preferred. Fixing now!